### PR TITLE
Clean up boards/shields in cmake and west

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Remove the boards and shields lists from the 'usage' target output.
That might have been readable at some point long ago in Zephyr's
history, when only a few boards were available, but right now it's
obscuring the high level targets we really want 'usage' to print.

Instead, add 'boards' and 'shields' targets which the user can run to
get those lists, and reference them from the 'usage' output. This
makes 'usage' squintable again. We use a new list_boards.py script
from the 'boards' target.

Reference the 'help' target from 'usage' as well, and drop the
recommendation that people run '--target help' from the 'west build
--help' output for the 'west build --target' option. The canonical
place to look is 'usage' now.

Use the new list_boards.py code from 'west boards' as well, which
allows us to add the board's directory as a format string key, in
addition to its name and architecture.

Keep west-completion.bash up to date. While doing that, I noticed that
a bunch of references to this file refer to a stale location, so fix
those too.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
The new `usage` output is:

```
Cleaning targets:
  clean     - Remove most generated files but keep configuration and backup files
  pristine  - Remove all files in the build directory

Kconfig targets:
  menuconfig - Update .config using a console-based interface
  guiconfig  - Update .config using a graphical interface

Other generic targets:
  all          - Build a zephyr application
  run          - Build a zephyr application and run it if the board supports emulation
  flash        - Run "west flash"
  debug        - Run "west debug"
  debugserver  - Run "west debugserver" (or start GDB server on port 1234 for QEMU targets)
  attach       - Run "west attach"
  ram_report   - Build and create RAM usage report
  rom_report   - Build and create ROM usage report
  boards       - Display supported boards
  shields      - Display supported shields
  usage        - Display this text
  help         - Display all build system targets

Build flags:

  ninja -v [targets] verbose build
  cmake -DW=n   Enable extra gcc checks, n=1,2,3 where
   1: warnings which may be relevant and do not occur too often
   2: warnings which occur quite often but may still be relevant
   3: more obscure warnings, can most likely be ignored
   Multiple levels can be combined with W=12 or W=123
```

**Comment by erwango** [comment]
@mbolivar-nordic Sorry for this basic question, but is there something to be done to see the new `usage` output you mention?
With this PR:
- I'm able to see that boards list is no more thrown in the west output when a wrong target name is provided
- `west boards` provide the list of boards
But:
- I don't get the `usage` output above
- `west shields` doesn't appear to be available:

```
$ west shields
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: error: argument <command>: invalid choice: 'shields' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate', 'completion', 'boards', 'build', 'sign', 'flash', 'debug', 'debugserver', 'attach', 'zephyr-export')
```
Did I missed something?

**Comment by utzig** [comment]
> * `west shields` doesn't appear to be available:

```
west build -t usage
```

> $ west shields

```
west build -t shields
```

**Review by utzig — APPROVED** [review]
LGTM. Nice improvement!

File: cmake/usage/CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,18 @@
...
+set(sorted_shield_list ${SHIELD_LIST})
+list(SORT sorted_shield_list)
+set(sorted_shield_cmds)
+foreach(shield ${sorted_shield_list})
+  list(APPEND sorted_shield_cmds COMMAND ${CMAKE_COMMAND} -E echo "  ${shield}")
+endforeach()
+add_custom_target(shields ${sorted_shield_cmds})
+
+include (${ZEPHYR_BASE}/cmake/boards.cmake)
+add_custom_target(boards ${list_boards_commands} USES_TERMINAL)
```
Comment (tejlmand) [inline cmake/usage/CMakeLists.txt:16]: I think creation of this target should happen directly inside: `cmake/boards.cmake`

File: cmake/usage/CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,18 @@
...
 string(REPLACE ";" " " SHIELD_LIST_SPACE_SEPARATED "${SHIELD_LIST}")
 
+string(REPLACE " " ";" SHIELD_LIST "${SHIELD_LIST_SPACE_SEPARATED}")
+set(sorted_shield_list ${SHIELD_LIST})
+list(SORT sorted_shield_list)
+set(sorted_shield_cmds)
+foreach(shield ${sorted_shield_list})
+  list(APPEND sorted_shield_cmds COMMAND ${CMAKE_COMMAND} -E echo "  ${shield}")
+endforeach()
+add_custom_target(shields ${sorted_shield_cmds})
```
Comment (tejlmand) [inline cmake/usage/CMakeLists.txt:13]: This is a bit short to place in a `shields.cmake` but I don't feel it belongs in this file.

So either in `shields.cmake` or alternatively as part of boards.cmake.

**Review by tejlmand — CHANGES_REQUESTED** [review]
If user specifies a wrong board, the current behavior is to print `usage`, which is one reason for printing the board as part of usage:
Today, if user does:
```
$ cmake -DBOARD=bad_board ...

No board named 'bad_board' found
....
see usage:
Cleaning targets:
...

Supported Boards:

  To generate project files for one of the supported boards below, run:

  $ cmake -DBOARD=<BOARD NAME> [-DSHIELD=<SHIELD NAME>] -Bpath/to/build_dir -Hpath/to/source_dir

  or

  $ export BOARD=<BOARD NAME>
  $ export SHIELD=<SHIELD NAME> #optional
  $ cmake -Bpath/to/build_dir -Hpath/to/source_dir

  arc:
    em_starterkit
    em_starterkit_em11d
    em_starterkit_em7d
    em_starterkit_em7d_v22
    emsdp
    emsdp_em4
    emsdp_em5d
<rest of boards>
```

With this PR:
```
No board named 'bad_board' found
...
see usage:
Cleaning targets:
  clean     - Remove most generated files but keep configuration and backup files
  pristine  - Remove all files in the build directory

Kconfig targets:
  menuconfig - Update .config using a console-based interface
  guiconfig  - Update .config using a graphical interface

Other generic targets:
<No board list printed>
```

I think that if user specifies a wrong board, then it is actually good to provide the board list, as that will be needed in order to find the right board.

But maybe it can be done in a more clever way.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Seems `zephyrbot` is re-requesting reviews.
Please see existing review here: https://github.com/zephyrproject-rtos/zephyr/pull/30572#pullrequestreview-550180083

File: cmake/usage/CMakeLists.txt (hunk 1 above)
Comment (mbolivar-nordic) [reply cmake/usage/CMakeLists.txt:16]: > I think creation of this target should happen directly inside: `cmake/boards.cmake`

add_custom_target() is not available in script mode, and boards.cmake still supports that.

File: cmake/usage/CMakeLists.txt (hunk 2 above)
Comment (mbolivar-nordic) [reply cmake/usage/CMakeLists.txt:13]: > So either in `shields.cmake` or alternatively as part of boards.cmake.

OK, I will add a 'shields.cmake'.

**Comment by mbolivar-nordic** [comment]
> ```
> 
> ```
> 
> 
> I think that if user specifies a wrong board, then it is actually good to provide the board list, as that will be needed in order to find the right board.
> 
> But maybe it can be done in a more clever way.

Done, and with shields list also.

Thanks for review, comments addressed.

File: cmake/app/boilerplate.cmake [hunk 3]
```diff
@@ -357,19 +357,37 @@ foreach(root ${BOARD_ROOT})
...
   endforeach()
-  print_usage()
+  message("Please choose from among the following shields:")
+  string(REPLACE ";" " " SHIELD_LIST_SPACE_SEPARATED "${SHIELD_LIST}")
+  execute_process(
+    COMMAND
+    ${CMAKE_COMMAND}
+    -DZEPHYR_BASE=${ZEPHYR_BASE}
+    -DSHIELD_LIST_SPACE_SEPARATED=${SHIELD_LIST_SPACE_SEPARATED}
+    -P ${ZEPHYR_BASE}/cmake/shields.cmake
```
Comment (tejlmand) [inline cmake/app/boilerplate.cmake:387]:
```suggestion
git add cmake/shields.cmake
git commit --amend
git push -f https://github.com/mbolivar-nordic/zephyr HEAD:drop-boards-from-usage
```
:wink:

File: cmake/usage/CMakeLists.txt (hunk 1 above)
Comment (tejlmand) [reply cmake/usage/CMakeLists.txt:11]: good point. :+1:

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please add `shields.cmake`.

File: cmake/app/boilerplate.cmake (hunk 3 above)
Comment (mbolivar-nordic) [reply cmake/app/boilerplate.cmake:387]: > wink

Thanks for the pro tip! Now done.

File: cmake/app/boilerplate.cmake [hunk 4]
```diff
@@ -357,19 +357,37 @@ foreach(root ${BOARD_ROOT})
...
+  message(FATAL_ERROR "Invalid BOARD; see above.")
 endif()
 
 if(DEFINED SHIELD AND NOT (SHIELD-NOTFOUND STREQUAL ""))
   foreach (s ${SHIELD-NOTFOUND})
     message("No shield named '${s}' found")
   endforeach()
-  print_usage()
+  message("Please choose from among the following shields:")
+  string(REPLACE ";" " " SHIELD_LIST_SPACE_SEPARATED "${SHIELD_LIST}")
```
Comment (tejlmand) [inline cmake/app/boilerplate.cmake:381]: You can pass a list directly to CMake.
But you would need to escape the `;` when going through `execute_process()`, which can be done as:
```suggestion
  string(REPLACE ";" "\\;" SHIELD_LIST_ESCAPED "${SHIELD_LIST}")
```

File: cmake/app/boilerplate.cmake [hunk 5]
```diff
@@ -357,19 +357,37 @@ foreach(root ${BOARD_ROOT})
...
+  execute_process(
+    COMMAND
+    ${CMAKE_COMMAND}
+    -DZEPHYR_BASE=${ZEPHYR_BASE}
+    -DSHIELD_LIST_SPACE_SEPARATED=${SHIELD_LIST_SPACE_SEPARATED}
```
Comment (tejlmand) [inline cmake/app/boilerplate.cmake:386]: escaping the list, means you can pass it directly as a list.
```suggestion
    -DSHIELD_LIST=${SHIELD_LIST_ESCAPED} 
```

File: cmake/shields.cmake [hunk 6]
```diff
@@ -0,0 +1,19 @@
+if(CMAKE_SCRIPT_MODE_FILE AND NOT CMAKE_PARENT_LIST_FILE)
+  # This file was invoked as a script directly with -P:
+  # cmake -P shields.cmake
+  #
+  # Unlike boards.cmake, this takes no OUTPUT_FILE option, but
+  # SHIELD_LIST_SPACE_SEPARATED is required.
+  string(REPLACE " " ";" SHIELD_LIST "${SHIELD_LIST_SPACE_SEPARATED}")
```
Comment (tejlmand) [inline cmake/shields.cmake:7]: which means you can avoid turning the space separated list into a CMake list, as it is already a CMake list.
Just use the list, as a list. That's the pro-tip :wink: 

```suggestion
```

**Review by tejlmand — COMMENTED** [review]
LGTM, just a minor suggestion.

File: cmake/shields.cmake (hunk 6 above)
Comment (mbolivar-nordic) [reply cmake/shields.cmake:7]: >  Just use the list, as a list. That's the pro-tip wink

Escaping the semicolons was what I was missing. Thanks.

**Comment by mbolivar-nordic** [comment]
Rebased and applied your suggestions @tejlmand . Thanks for the pointers.

File: cmake/boards.cmake [hunk 7]
```diff
@@ -1,91 +1,21 @@
 # SPDX-License-Identifier: Apache-2.0
 
-# List all architectures, export the list in list_var
-function(list_archs list_var)
+string(REPLACE ";" " " BOARD_ROOT_SPACE_SEPARATED "${BOARD_ROOT}")
```
Comment (tejlmand) [inline cmake/boards.cmake:3]: I think this can also be removed.
```suggestion
```

**Review by tejlmand — APPROVED** [review]
Approved, as code looks good.
Thanks for this great improvement in the user experience.

Please remove the obsolete line before merging.

**Pull request merged**